README
This program is similar to ping(1), but tests connectivity at the
link layer (layer-2) instead of the network layer (layer-3).

HPUX has linkloop support for ages, and lately I googled a little
and found that link-layer connectivity loopback testing is part
of the IEEE 802.2LLC standard.

So I went and written:
	- linkloop for Linux.
	- linkloop_reply as a user space reply program. Used for
	  testing against Linux until the Linux kernel has native
	  support for this protocol (in my TODO :-)
	- lanscan is a small utility to show local interfaces
	  with their mac addresses. Yes I know 'ip link' gives
	  the same information, I just wanted to play with some
	  ioctls :-)

To get me started I grabbed some examples from spak
	http://www.xenos.net/software/spak/
(Which wasn't compiling cleanly on my machine BTW).

It was tested against HPUX-11.0 and Linux (when running linkloop_reply)
both ways.

My linkloop accepts MAC address in:
	- HPUX notation (0x......)
	- Colon separated (00:BC:.....)
	- A name from /etc/ethers


NEWS
* Changes from 1.0.0 by Dominique Domet de Mont <Dominique.Domet-de-Mont@hp.com>
  - linkloop_reply replies only on the interface having received the packet, instead of all.

* Changes from 0.0.2 by Dominique Domet de Mont <Dominique.Domet-de-Mont@hp.com>
  - MAC address output format changed to 0x%02X%02X%02X%02X%02X%02X to mimic HP UX ouput
  - do not stop listening when receiving an unexpected packet
  - lanscan options to support various display options
  - linkloop_reply able to listen several ports passed as parameter instead of just eth0
  - /etc/init.d/linkloopd service shell script taking care of starting and stopping linkloop_reply 
  
* Changes from 0.0.1
  - Updated man page documenting all options
  - Better validation and reporting of received packets


common.c
/*
 * Written by Oron Peled <oron@actcom.co.il>
 * 
 * Modified by Dominique Domet de Mont, 2007, Dominique.Domet-de-Mont@hp.com
 * Copyright (C) 2007 Hewlett-Packard Development Company, L.P.
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 */

#include "config.h"

#include <netinet/in.h>		/* for htons(3)			*/
#include <linux/if_packet.h>

#if HAVE_ETHER_HOSTTON
#include <netinet/ether.h>	/* for ether_hostton(3)		*/
#endif

#include <net/if.h>		/* for IF_NAMESIZE, IFHWADDRLEN */
#include <sys/ioctl.h>		/* for SIOCGIFHWADDR		*/
#include <stdio.h>
#include <string.h>     	/* for memcmp(3) bzero(3)	*/
#include <stdlib.h>		/* for exit(3)			*/
#include <errno.h>
#include <assert.h>

#include "linkloop.h"

/* These are used to define different data structs... */
#if IFHWADDRLEN != ETH_ALEN
#error "SOMETHING IS VERY FISHY: IFHWADDRLEN != ETH_ALEN"
#endif

int debug_flag = 0;

void dump_packet(struct llc_packet *pack) {
	int i;
	u_int8_t *p = (u_int8_t *)pack;
	size_t len = ntohs(pack->eth_hdr.ether_type);

	printf("PACKET DUMP: data size=%d (0x%x)", len, len);
	for (i = 0; i < len; i++, p++) {
		if ((i % 16) == 0)
			printf("\n%04x\t", i);
		printf("%02x ", (unsigned)*p);
	}
	printf("\nEND PACKET DUMP\n");
}

char *mac2str(u_int8_t *s) {
	static char buf[3*ETH_ALEN];

	sprintf(buf, "0x%02X%02X%02X%02X%02X%02X",
		s[0], s[1], s[2], s[3], s[4], s[5]);
	return buf;
}

int parse_address(u_int8_t mac[], const char *str) {
	unsigned a, b, c, d, e, f;
	struct ether_addr ea;

	if (sscanf(str,"%x:%x:%x:%x:%x:%x", &a, &b, &c, &d, &e, &f) == 6) {
		/* A colon separated notation */
		mac[0] = (unsigned char) a;
		mac[1] = (unsigned char) b;
		mac[2] = (unsigned char) c;
		mac[3] = (unsigned char) d;
		mac[4] = (unsigned char) e;
		mac[5] = (unsigned char) f;
	} else if (sscanf(str, "0x%02x%02x%02x%02x%02x%02x", &a, &b, &c, &d, &e, &f) == 6) {
		/* Hexadecimal notation (like HPUX) */
		mac[0] = (unsigned char) a;
		mac[1] = (unsigned char) b;
		mac[2] = (unsigned char) c;
		mac[3] = (unsigned char) d;
		mac[4] = (unsigned char) e;
		mac[5] = (unsigned char) f;
#if HAVE_ETHER_HOSTTON
	} else if (ether_hostton(str, &ea) == 0) {
		/* A name from /etc/ethers */
		memcpy(mac, ea.ether_addr_octet, 8);
#endif
	} else
		return 0;
	return 1;
}

void get_hwaddr(int sock, const char iface[], int *ifindex, u_int8_t mac[]) {
	struct ifreq ifr;

	strncpy(ifr.ifr_name, iface, IFNAMSIZ);
	if (ioctl(sock, SIOCGIFHWADDR, &ifr) < 0) {
		perror("ioctl(SIOCGIFHWADDR)");
		exit(1);
	}
	memcpy(mac, ifr.ifr_hwaddr.sa_data, IFHWADDRLEN);

	if (ioctl(sock, SIOCGIFINDEX, &ifr) < 0) {
		perror("ioctl(SIOCGIFINDEX)");
		exit(1);
	}
	*ifindex = ifr.ifr_ifindex;
}

void mk_test_packet(struct llc_packet *pack, struct sockaddr_ll *sll,
		    const u_int8_t src_mac[], int src_ifindex,
		    const u_int8_t dst_mac[], size_t len, int response) {
	int i;

	bzero(sll, sizeof(*sll));
	sll->sll_family = AF_PACKET;
	sll->sll_protocol = htons(ETH_P_802_2);
	sll->sll_ifindex = src_ifindex;
	sll->sll_halen = ETH_ALEN;
	memcpy(sll->sll_addr, dst_mac, ETH_ALEN);

	assert(len <= ETH_DATA_LEN);			/* 0x05DC == 1500 */
	memcpy(pack->eth_hdr.ether_dhost, dst_mac, IFHWADDRLEN);
	memcpy(pack->eth_hdr.ether_shost, src_mac, IFHWADDRLEN);
	pack->eth_hdr.ether_type = htons(len);
	pack->llc.dsap = (response) ? 0x80 : 0x00;
	pack->llc.ssap = (response) ? 0x01 : 0x80;	/* XNS? */
	pack->llc.ctrl = TEST_CMD;			/* TEST */

	for (i = 0; i < DATA_SIZE(len); i++)
		pack->data[i] = i;
}

void send_packet(int sock, struct llc_packet *pack, struct sockaddr_ll *sll) {
	int ret;

	/* Send the packet */
	ret = sendto(sock, pack, FRAME_SIZE(pack), 0, (struct sockaddr *)sll, sizeof(*sll));
	if (ret == -1) {
		perror("sendto");
		exit(1);
	}
	if (ret != FRAME_SIZE(pack))
		fprintf(stderr, "Warning: Incomplete packet sent\n");

	if (debug_flag) {
		printf("sent TEST packet (%d bytes) to %s\n",
			ntohs(pack->eth_hdr.ether_type), mac2str(pack->eth_hdr.ether_dhost));
	}
}

int recv_packet(int sock, struct llc_packet *pack) {
	struct sockaddr_ll sll = {0};
	socklen_t len;
	int ret;

	len = sizeof(sll);
	ret = recvfrom(sock, pack, sizeof(*pack), 0, (struct sockaddr *)&sll, &len);
	if (ret == -1) {
		if(errno == EINTR)	/* We have a timeout */
			return 0;
		perror("recvfrom");
		exit(1);
	}

	if ((pack->llc.ctrl & TEST_CMD) != TEST_CMD) {
		fprintf(stderr, "got unexpected packet\n");
		dump_packet(pack);
		/* continue anyway ! exit(1);*/
	}
	ret = ntohs(pack->eth_hdr.ether_type);
	if (debug_flag)
		printf("received TEST packet (%d bytes) from %s\n", ret, mac2str(pack->eth_hdr.ether_shost));

	return ret;
}


lanscan.c
/*
 * Written by Oron Peled <oron@actcom.co.il>
 *
 * Modified by Dominique Domet de Mont, 2007, Dominique.Domet-de-Mont@hp.com
 * Copyright (C) 2007 Hewlett-Packard Development Company, L.P.
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include "config.h"
#include "linkloop.h"

/* Linkloop configuration */
static struct lanscan {
	char displayHeader; 	/* display header before listing interfaces */
	char displayMac; 	/* display the MAC address */
	char displayNames;	/* display interface names */
	char displayState;	/* display interface states */
	char displayOneline;	/* display one line summary */

} ls = {
	.displayHeader = 1,
	.displayMac = 1,
	.displayNames = 1,
	.displayState = 1,
	.displayOneline = 0
};

static struct if_flag {
	short int flag;
	char *flag_name;
} if_flags[] = {
	{ IFF_UP, "UP" },
	{ IFF_BROADCAST, "BROADCAST" },
	{ IFF_DEBUG, "DEBUG" },
	{ IFF_LOOPBACK, "LOOPBACK" },
	{ IFF_POINTOPOINT, "POINTOPOINT" },
	{ IFF_NOTRAILERS, "NOTRAILERS" },
	{ IFF_RUNNING, "RUNNING" },
	{ IFF_NOARP, "NOARP" },
	{ IFF_PROMISC, "PROMISC" },
	{ IFF_ALLMULTI, "ALLMULTI" },
	{ IFF_MASTER, "MASTER" },
	{ IFF_SLAVE, "SLAVE" },
	{ IFF_MULTICAST, "MULTICAST" },
	{ IFF_PORTSEL, "PORTSEL" },
	{ IFF_AUTOMEDIA, "AUTOMEDIA" },
};

static short int get_ifflags(int sock, const char name[]) {
	struct ifreq ifr;

	strncpy(ifr.ifr_name, name, IFNAMSIZ);
	if (ioctl(sock, SIOCGIFFLAGS, &ifr) < 0) {
		perror("ioctl(SIOCGIFFLAGS)");
		exit(1);
	}
	return ifr.ifr_flags;
}

static void usage(const char *program) {
	fprintf(stderr, "Usage: %s -[option...]\n"
		"\t-h		Print this help text\n"
		"\t-?		Print this help text\n"
		"\t-i<face>	Print network interface names\n"
		"\t-a<mAc>	Print network interface MAC address\n"
		"\t-s<tate>	Print network interface state\n"
		"\t-o<ne>	Print one line summary, interface names only\n"
		, program
	);
	exit(1);
}

int main(int argc, char *argv[]) {
	struct ifconf ifc;
	struct ifreq ifr_x[MAX_IFACES];
	int sock, err, ifindex, i;
	u_int8_t myMAC[IFHWADDRLEN];
	short int flags;

	/* check options if any */
	if(argc > 1) {
		/* unsupported options or help */
		if(strpbrk(argv[1], "aiso") == NULL) {
			usage(argv[0]);
		}
		
		/* supported options provided: turn the header display off */
		ls.displayHeader = 0;
		if (strchr(argv[1], 'i') == NULL)
			ls.displayNames = 0;
		if (strchr(argv[1], 'a') == NULL)
			ls.displayMac = 0;
		if (strchr(argv[1], 's') == NULL)
			ls.displayState = 0;
		if (strchr(argv[1], 'o')) {
			ls.displayOneline = 1;
			ls.displayState = ls.displayMac = ls.displayNames = 0;
		}
	}
	
	if ((sock = socket(PF_PACKET, SOCK_PACKET, 0)) < 0) {
		perror("socket");
		exit(1);
	}

	ifc.ifc_len = MAX_IFACES * sizeof(struct ifreq);
	ifc.ifc_req = ifr_x;
	if ((err = ioctl(sock, SIOCGIFCONF, &ifc)) < 0) {
		perror("ioctl");
		exit(1);
	}
	if (ls.displayHeader)
		printf("retrieved info for %i interface(s)\n", ifc.ifc_len / sizeof(struct ifreq));

	for (err = 0; err < ifc.ifc_len / sizeof(struct ifreq); err++) {
		flags = get_ifflags(sock, ifr_x[err].ifr_name);
		get_hwaddr(sock, ifr_x[err].ifr_name, &ifindex, myMAC);
		
		/* Discard local loopback port */
		if (strcmp(ifr_x[err].ifr_name, "lo") == 0)
			continue;
			
		/* Discard alias ports, ie name including a column ':' */
		if (strchr(ifr_x[err].ifr_name, ':'))
			continue;
			
		if (ls.displayOneline) {
			printf("%s\t", ifr_x[err].ifr_name);
			continue;
		}
		
		if (ls.displayMac)
			printf("%s\t", mac2str(myMAC));

		/* here, in order to mimic the HP UX output, we print the interface name,
		 * interface physical point of attachement, physical point ID
		 * as the same "interface name"
		 */
		if (ls.displayNames) {
			printf("%s %s %s\t", ifr_x[err].ifr_name, ifr_x[err].ifr_name,
					     ifr_x[err].ifr_name);
		}

		if (ls.displayState) {
			putchar('<');
			for (i = 0; i < sizeof(if_flags)/sizeof(if_flags[0]); i++) {
				if (flags & if_flags[i].flag)
					printf("%s ", if_flags[i].flag_name);
			}
			putchar('>');
		}
		putchar('\n');
	}

	return EXIT_SUCCESS;
}


linkloop.h
/*
 * Written by Oron Peled <oron@actcom.co.il>
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 */

#ifndef	LINKLOOP_H
#define	LINKLOOP_H

#include <linux/if_packet.h>
#include <netinet/if_ether.h>

#define DATA_SIZE(_pack_size)	(_pack_size - sizeof(struct llc))
#define	MAX_DATA_SIZE		DATA_SIZE(ETH_DATA_LEN)	/* This is what HPUX sends and responds to */
#define FRAME_SIZE(_pack)	(ntohs((_pack)->eth_hdr.ether_type) + sizeof(struct ether_header))

#define	MAX_IFACES	20

/* Describe an 802.2LLC packet */
struct llc {
	unsigned char dsap;
	unsigned char ssap;
	unsigned char ctrl;
} __attribute__ ((__packed__));

struct llc_packet {
	struct ether_header eth_hdr;
	struct llc llc;
	/* Some test data */
	unsigned char data[MAX_DATA_SIZE];
} __attribute__ ((__packed__));

#define	TEST_CMD	0xE3	/* From 802.2LLC (HPUX sends 0xF3) */

extern int debug_flag;

void dump_packet(struct llc_packet *pack);
char *mac2str(u_int8_t *s);
int parse_address(u_int8_t mac[], const char *str);
void get_hwaddr(int sock, const char name[], int *ifindex, u_int8_t mac[]);
void mk_test_packet(struct llc_packet *pack, struct sockaddr_ll *sll,
		    const u_int8_t src_mac[], int src_ifindex,
		    const u_int8_t dst_mac[], size_t len, int response);
void send_packet(int sock, struct llc_packet *pack, struct sockaddr_ll *sll);
int recv_packet(int sock, struct llc_packet *pack);

#endif	/* LINKLOOP_H */


linkloop_reply.c
/*
 * Written by Oron Peled <oron@actcom.co.il>
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 */
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <netinet/if_ether.h>
#include <linux/if_packet.h>
#include <net/if.h>
#include <sys/select.h>
#include "config.h"
#include "linkloop.h"

/*
 * This program runs as a deamon, listening at the llc level on network interfaces 
 * passed on the command line, and responding by sending back a test packet.
 */

/* Linkloop reply configuration */
static struct linkloop_reply {
	const char *program;

	int ifindex_listen[MAX_IFACES];
	u_int8_t mac_listen[MAX_IFACES][IFHWADDRLEN];
	u_int8_t src_mac[IFHWADDRLEN];
	u_int8_t dst_mac[IFHWADDRLEN];
} llr;
 
int main(int argc, char *argv[]) {
	struct llc_packet spack;
	struct llc_packet rpack;
	struct sockaddr_ll sll;
	int nif = argc - 1;	/* number of interfaces to listen */
	size_t len;
	int sock;
	int i;
	
	llr.program = argv[0];

	if (nif == 0 || nif > MAX_IFACES) {
		/* one parameter expected at least: the lan interface name to listen */
		fprintf(stderr, "Expecting a list of one to %d interfaces to listen; eg: %s eth0 eth3\n",
			MAX_IFACES, llr.program);
		return 1;
	}

	/* Open the socket */
	if ((sock = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_802_2))) == -1) {
		perror("socket");
		return 1;
	}
	
	/* Getting mac addresses for all listened interfaces */
	for(i = 0; i < nif; ++i) {
		get_hwaddr(sock, argv[i+1], &llr.ifindex_listen[i], llr.mac_listen[i]);
	}

	/* listen and reply forever */
	do {
		len = recv_packet(sock, &rpack);
		memcpy(llr.src_mac, rpack.eth_hdr.ether_shost, IFHWADDRLEN);
		memcpy(llr.dst_mac, rpack.eth_hdr.ether_dhost, IFHWADDRLEN);

		/* check against listened interfaces */
		for(i = 0; i < nif; ++i)
		{
			/* check if this packet has been received on a listened interface */
			if(memcmp(llr.mac_listen[i], llr.dst_mac, IFHWADDRLEN))
				continue;
				
			/* return a test packet to the sender */
			printf("Received packet on %s\n", argv[i+1]);
			mk_test_packet(&spack, &sll, llr.dst_mac, llr.ifindex_listen[i], llr.src_mac, len, 1);
			send_packet(sock, &spack, &sll);
		}
	} while(1);

	return 0;
}


linkloop.c
/*
 * linkloop: Loopback testing for layer-2 connectivity
 *           via 802.2LLC TEST message.
 *
 * Tested against HPUX-11.0
 *
 * Written by Oron Peled <oron@actcom.co.il>
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <getopt.h>
#include <unistd.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <linux/if_packet.h>
#include <net/if.h>
#include <arpa/inet.h>

#include "config.h"
#include "linkloop.h"

/* Linkloop configuration */
static struct linkloop {
	const char *program;

	size_t pack_size;
	int timeout;
	int retries;

	const char *src_iface;
	int src_ifindex;
	u_int8_t src_mac[IFHWADDRLEN];
	const char *dst_mac_str;
	u_int8_t dst_mac[IFHWADDRLEN];

	unsigned int total_sent;
	unsigned int total_good;
	unsigned int total_timeout;
	unsigned int total_bad;
} ll = {
	.pack_size = ETH_DATA_LEN,	/* 0x05DC == 1500 */
	.timeout = 2,
	.retries = 1,
	.src_iface = "eth0"
};

static void usage(void) {
	fprintf(stderr, "Usage: %s [option...] mac_addr\n"
		"\t-d		Debug\n"
		"\t-i<iface>	Network interface\n"
		"\t-t<timeout>	Timeout between packets\n"
		"\t-n<retries>	Number of retries\n"
		"\t-s<size>	Packet size in bytes\n"
		, ll.program
	);
	exit(1);
}

static void handle_options(int argc, char * const argv[]) {
	int c;

	ll.program = argv[0];

	while ((c = getopt(argc, argv, "di:t:n:s:")) != -1) {
		switch(c) {
		case 'd':
			debug_flag = 1;
			break;
		case 'i':
			ll.src_iface = optarg;
			break;
		case 't':
			ll.timeout = strtol(optarg, NULL, 0);
			break;
		case 'n':
			ll.retries = strtol(optarg, NULL, 0);
			break;
		case 's':
			ll.pack_size = strtol(optarg, NULL, 0);
			if(ll.pack_size > ETH_DATA_LEN) {
				fprintf(stderr, "%s: Illegal size specified %d > %d (ETH_DATA_LEN)\n",
					ll.program, ll.pack_size, ETH_DATA_LEN);
				exit(1);
			}
			break;
		default:
			fprintf(stderr, "%s: unknown option code 0x%x\n", ll.program, c);
		case '?':
			usage();
		}
	}

	if (debug_flag) {
		printf("interface=%s timeout=%d num=%d size=%d\n",
			ll.src_iface, ll.timeout, ll.retries, ll.pack_size);
	}

	if (optind != argc - 1) {
		fprintf(stderr, "%s: missing target address\n", ll.program);
		usage();
	}
	ll.dst_mac_str = argv[optind];
}

static void sig_alarm(int signo) {}

static void set_sighandlers(void) {
	struct sigaction sa;

	sa.sa_handler = sig_alarm;
	sa.sa_flags = 0;

	sigfillset(&sa.sa_mask);	/* Block spurious signals during handler */
	sigdelset(&sa.sa_mask, SIGINT);	/* But normal signals should kill us */
	sigdelset(&sa.sa_mask, SIGQUIT);
	sigdelset(&sa.sa_mask, SIGTERM);

	if (sigaction(SIGALRM, &sa, NULL) < 0) {
		perror("sigaction");
		exit(1);
	}
}

static int linkloop(int sock) {
	struct llc_packet spack;
	struct llc_packet rpack;
	struct sockaddr_ll sll;
	int ret;

	mk_test_packet(&spack, &sll, ll.src_mac, ll.src_ifindex, ll.dst_mac, ll.pack_size, 0);
	if (alarm(ll.timeout) < 0) {
		perror("alarm");
		exit(1);
	}

	send_packet(sock, &spack, &sll);
	ll.total_sent++;
	ret = recv_packet(sock, &rpack);
	if (ret == 0) { /* timeout */
		fprintf(stderr, "  ** TIMEOUT (%d seconds)\n", ll.timeout);
		ll.total_timeout++;
		return 0;
	}

	if (spack.eth_hdr.ether_type != rpack.eth_hdr.ether_type) {
		ll.total_bad++;
		printf("  ** BAD RECEIVED LENGTH = %d\n", rpack.eth_hdr.ether_type);
		return 0;
	}

	if (memcmp(spack.eth_hdr.ether_dhost, rpack.eth_hdr.ether_shost, IFHWADDRLEN) != 0 &&
	    memcmp(spack.eth_hdr.ether_shost, rpack.eth_hdr.ether_shost, IFHWADDRLEN) != 0) {
		ll.total_bad++;
		printf("  ** ROGUE RESPONDER: received from %s\n",
			mac2str(rpack.eth_hdr.ether_shost));
		return 0;
	}

	if (memcmp(spack.data, rpack.data, DATA_SIZE(ll.pack_size)) != 0) {
		ll.total_bad++;
		printf("  ** BAD RESPONSE\n");
		dump_packet(&rpack);
		return 0;
	}

	ll.total_good++;
	return 1;
}

int main(int argc, char *argv[]) {
	int sock, i;

	handle_options(argc, argv);

	if (!parse_address(ll.dst_mac, ll.dst_mac_str)) {
		fprintf(stderr, "%s: bad DST address - %s\n", ll.program, ll.dst_mac_str);
		return 1;
	}

	printf("Link connectivity to LAN station: %s (HW addr %s)\n", ll.dst_mac_str, mac2str(ll.dst_mac));

	if ((sock = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_802_2))) == -1) {
		perror("socket");
		return 1;
	}

	if (debug_flag)
		fprintf(stderr, "Getting MAC address of interface '%s'\n", ll.src_iface);
	get_hwaddr(sock, ll.src_iface, &ll.src_ifindex, ll.src_mac);
	if (debug_flag)
		fprintf(stderr, "Testing via %s (HW addr %s)\n", ll.src_iface, mac2str(ll.src_mac));

	set_sighandlers();
	for (i = 0; i < ll.retries; i++) {
		(void)linkloop(sock);

		if (debug_flag)
			printf("Retry %d...\n", i);
	}

	if (ll.total_sent == ll.total_good) {
		printf("  -- OK -- %d packets\n", ll.total_sent);
		return 0;
	} else if (ll.total_sent == ll.total_timeout) {
		printf("  -- NO RESPONSE --\n");
	} else {
		printf("  -- %u packets transmitted, %u received, %u timed out, %u bad --\n",
			ll.total_sent, ll.total_good, ll.total_timeout, ll.total_bad);
	}

	return 1;
}
